l1-contracts/contracts/bridge/asset-tracker/AssetTracker.sol

```diff
@@ -294,80 +435,83 @@ contract AssetTracker is IAssetTracker, Ownable2StepUpgradeable, AssetHandlerMod
         require(_finalizeWithdrawalParams.l2Sender == L2_ASSET_TRACKER_ADDR, InvalidSender());
 
         // solhint-disable-next-line no-unused-vars
-        (
-            uint256 chainId,
-            bytes32 assetId,
-            uint256 amount,
-            uint256 migrationNumber,
-            bool isL1ToGateway
-        ) = _decodeTokenBalanceMigrationData(_finalizeWithdrawalParams.message);
-
-        if (!isL1ToGateway) {
-            // here another settlement layer might frontrun
-            require(BRIDGE_HUB.whitelistedSettlementLayers(_finalizeWithdrawalParams.chainId), InvalidChainId());
-        }
 
-        uint256 currentSettlementLayer = BRIDGE_HUB.settlementLayer(chainId);
+        TokenBalanceMigrationData memory data = abi.decode(
+            _finalizeWithdrawalParams.message,
+            (TokenBalanceMigrationData)
+        );
+        require(assetMigrationNumber[data.chainId][data.assetId] < data.migrationNumber, InvalidAssetId());
+
+        uint256 currentSettlementLayer = BRIDGE_HUB.settlementLayer(data.chainId);
         // require(_getMigrationNumber(chainId) == migrationNumber, InvalidMigrationNumber());
-        if (isL1ToGateway) {
+        if (data.isL1ToGateway) {
             require(currentSettlementLayer != block.chainid, NotMigratedChain());
-            require(chainId == _finalizeWithdrawalParams.chainId, InvalidChainId());
-            require(assetSettlementLayer[assetId] == block.chainid, InvalidAssetId());
-
-            assetSettlementLayer[assetId] = currentSettlementLayer;
-
-            chainBalance[chainId][assetId] -= amount;
-            chainBalance[currentSettlementLayer][assetId] += amount;
-
-            _sendConfirmToGateway({
-                _settlementLayer: currentSettlementLayer,
-                _chainId: chainId,
-                _assetId: assetId,
-                _amount: amount,
-                _isL1ToGateway: isL1ToGateway
-            });
-            _sendConfirmToL2(chainId, assetId, amount, migrationNumber);
+            require(data.chainId == _finalizeWithdrawalParams.chainId, InvalidWithdrawalChainId());
+            uint256 chainMigrationNumber = _getMigrationNumber(data.chainId);
+
+            // we check parity here to make sure that we migrated back to L1 from Gateway.
+            // In the future we might initalize chains on GW. So we subtract from chainMigrationNumber.
+            require(
+                chainMigrationNumber - ((assetMigrationNumber[data.chainId][data.assetId]) % 2) == 1,
+                InvalidMigrationNumber()
+            );
+
+            _ensureTokenIsRegistered(data.assetId, data.tokenOriginChainId);
```

Comment:
this function will overwrite `isMinterChain` based on untrusted data provided by the chain, this seems wrong

Reply:
I completely removed isMinterChain storage, now we do assetId checks and include the tokenOriginChainId